features/features.go

```diff
@@ -0,0 +1,112 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package features
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+)
+
+// SnapdFeature is a named feature that may be on or off.
+type SnapdFeature int
+
+const (
+	// Layouts controls availability of snap layouts.
+	Layouts SnapdFeature = iota
+	// ParallelInstances controls availability installing a snap multiple times.
+	ParallelInstances
+	// Hotplug controls availability of dynamically creating slots based on system hardware.
+	Hotplug
+	// SnapdSnap controls possibility of installing the snapd snap.
+	SnapdSnap
+	// PerUserMountNamespace controls the persistence of per-user mount namespaces.
+	PerUserMountNamespace
+)
+
+// KnownFeatures returns the list of all known features.
+func KnownFeatures() []SnapdFeature {
+	return []SnapdFeature{Layouts, ParallelInstances, Hotplug, SnapdSnap, PerUserMountNamespace}
+}
+
+// String returns the name of a snapd feature.
+func (f SnapdFeature) String() string {
+	// The constants here must be synchronized with cmd/libsnap-confine-private/feature.c
+	switch f {
+	case Layouts:
+		return "layouts"
+	case ParallelInstances:
+		return "parallel-instances"
+	case Hotplug:
+		return "hotplug"
+	case SnapdSnap:
+		return "snapd-snap"
+	case PerUserMountNamespace:
+		return "per-user-mount-namespace"
+	}
+	panic(fmt.Sprintf("unknown feature flag code %d", f))
+}
+
+// IsEnabledWhenUnset returns true if a feature is enabled when not set.
+//
+// A feature may be enabled or disabled with explicit state in snapd. If
+// explicit state is absent the effective value is the implicit default
+// computed by this function.
+func (f SnapdFeature) IsEnabledWhenUnset() bool {
+	switch f {
+	case Layouts:
+		return true
+	}
+	return false
+}
+
+// IsExported returns true if a feature is copied from snapd state to a feature file.
+//
+// Certain features are available outside of snapd internal state and visible as control
+// files in a dedicated directory. Such features can be queried for, via IsEnabled, outside
+// of snapd.
+func (f SnapdFeature) IsExported() bool {
+	switch f {
+	case PerUserMountNamespace:
+		return true
+	}
+	return false
+}
+
+// ControlFile returns the path of the file controlling the exported feature.
+//
+// Snapd considers the feature enabled if the file is present.
+// The contents of the file are not important.
+func (f SnapdFeature) ControlFile() string {
+	if !f.IsExported() {
+		panic(fmt.Sprintf("cannot compute the control file of feature %q because that feature is not exported", f))
```

Comment:
Same as `IsExported`, do we need to export this method? I think `IsEnabled` is all we need?

Reply:
This is used by the configuration manager and hear internally.